crates/turbopack-node/src/evaluate.rs

```diff
@@ -29,17 +32,67 @@ use crate::{
     bootstrap::NodeJsBootstrapAsset,
     embed_js::embed_file_path,
     emit, emit_package_json,
-    pool::{NodeJsPool, NodeJsPoolVc},
-    EvalJavaScriptIncomingMessage, EvalJavaScriptOutgoingMessage, StructuredError,
+    pool::{NodeJsOperation, NodeJsPool, NodeJsPoolVc},
+    StructuredError,
 };
 
+#[derive(Serialize)]
+#[serde(tag = "type", rename_all = "camelCase")]
+enum EvalJavaScriptOutgoingMessage<'a> {
+    #[serde(rename_all = "camelCase")]
+    Evaluate { args: Vec<&'a JsonValue> },
+}
+
+#[derive(Deserialize)]
+#[serde(tag = "type", rename_all = "camelCase")]
+enum EvalJavaScriptIncomingMessage {
+    FileDependency {
+        path: String,
+    },
+    BuildDependency {
+        path: String,
+    },
+    DirDependency {
+        path: String,
+        glob: String,
+    },
+    EmittedError {
+        severity: IssueSeverity,
+        error: StructuredError,
+    },
+    Value {
+        // TODO: There is like 3 levels of JSON string encoding that goes into returning a value,
+        // making it really inefficient.
+        data: String,
+    },
+    End {
+        data: Option<String>,
+    },
+    Error(StructuredError),
+}
+
+#[derive(Debug)]
+enum LoopResult {
+    Value(String),
+    End(Option<String>),
+    Error(StructuredError),
+}
+
+type EvaluationItem = Result<Bytes, StructuredError>;
+type JavaScriptStream = Stream<EvaluationItem>;
+
 #[turbo_tasks::value(shared)]
 #[derive(Clone, Debug)]
-pub enum JavaScriptValue {
-    Error,
-    Value(Rope),
-    // TODO, support stream in the future
-    Stream(#[turbo_tasks(trace_ignore)] Vec<u8>),
+pub enum JavaScriptEvaluation {
+    /// An Empty is only possible if the evaluation completed successfully, but
+    /// didn't send an error nor value.
+    Empty,
+    /// A Single is only possible if the evaluation returns either an error or
+    /// value, and never sends an intermediate value.
+    Single(EvaluationItem),
+    /// A Stream represents a series of intermediate values and followed by
+    /// either an error or a ending value. A stream is never empty.
+    Stream(#[turbo_tasks(trace_ignore, debug_ignore)] JavaScriptStream),
 }
```

Comment:
I think it would be easier to consume when it is always an stream.

```suggestion
pub enum JavaScriptEvaluation(#[turbo_tasks(trace_ignore, debug_ignore)] JavaScriptStream)
```

Maybe add `StreamExt::into_single_value()` which gets the only value from stream and throws when it's empty, more than one value or the stream errored.

Reply:
hmm?

``` diff
- let JavaScriptEvaluation::Single(Ok(val)) = &*config_value else {
+ let Some(val) = config_value.into_single().await? else {
```

Reply:
Currently the configs handle errors by defaulting, moving the unwrap into the `?` would prevent that.

Reply:
I figured out a return value that makes this palatable.

One side effect of this is that any code that handles either a single value or a stream of values must ensure they fully close the stream so the `NodeJsOperation` is released back to the pool. Eg, the `route` method can take receive either a single result for `Rewrite`, or a stream for `Middleware` responses. I can't call `into_single` for this, because I need the individual results for the middleware. For the rewrite response, I just called the `read.next().await` a single time. But that left the stream midway, and the process was never released to the pool. I had to add a bit of code to call into the stream one more time.